src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,535 @@
+      return;
+    }
+    Array<ObjectRef> res = dtype_convert_info.value();
+    MixedTypePolicy policy = static_cast<MixedTypePolicy>(Downcast<Integer>(res[0])->value);
+    if (policy == MIXED_PRECISION_ALWAYS) {
+      // require inputs to be fp16
+      RequireArgsToType(call_node->args, fp16_);
+    } else if (policy == MIXED_PRECISION_FOLLOW) {
+      // no requirement
+      auto it = var_dtype_map_.find(binding->var);
+      if (it == var_dtype_map_.end()) {
+        LOG(WARNING) << binding->value << " is not used in the program";
+        return;
+      }
+      NTensorStruct all_fp32 = NTensorStructFrom(binding->var, fp32_);
+      DataType to;
+      if (it->second.find(all_fp32) == it->second.end() || it->second.size() > 1) {
+        to = fp16_;
+      } else {
+        to = fp32_;
+      }
+      RequireArgsToType(call_node->args, to);
+    } else if (policy == MIXED_PRECISION_NEVER) {
+      // require inputs to be fp32
+      RequireArgsToType(call_node->args, fp32_);
+    } else {
+      LOG(FATAL) << "Unsupported MixedTypePolicy: " << policy;
+    }
+  }
+
+  // override the following methods to visit in backward order
+  void VisitExpr_(const SeqExprNode* op) final {
+    this->VisitSpan(op->span);
+    this->VisitExpr(op->body);
+    for (auto it = op->blocks.rbegin(); it != op->blocks.rend(); it++) {
+      this->VisitBindingBlock(*it);
+    }
+
+    if (auto* sinfo = op->struct_info_.as<StructInfoNode>()) {
+      this->VisitExprDepStructInfoField(GetRef<StructInfo>(sinfo));
+    }
+  }
+
+  void VisitBindingBlock_(const BindingBlockNode* block) {
+    for (auto it = block->bindings.rbegin(); it != block->bindings.rend(); it++) {
+      this->VisitBinding(*it);
+    }
+  }
+
+  void VisitBindingBlock_(const DataflowBlockNode* block) {
+    for (auto it = block->bindings.rbegin(); it != block->bindings.rend(); it++) {
+      this->VisitBinding(*it);
+    }
+  }
+
+  void VisitExpr_(const IfNode* op) final {
+    this->VisitSpan(op->span);
+    this->VisitExpr(op->true_branch);
+    this->VisitExpr(op->false_branch);
+    this->VisitExpr(op->cond);
+
+    if (auto* sinfo = op->struct_info_.as<StructInfoNode>()) {
+      this->VisitExprDepStructInfoField(GetRef<StructInfo>(sinfo));
+    }
+  }
+
+  DataType fp16_ = DataType(DataType::TypeCode::kFloat, 16, 1);
+  DataType fp32_ = DataType(DataType::TypeCode::kFloat, 32, 1);
+  DataType output_dtype_;
+};
+
+class ToMixedPrecisionRewriter : public ExprMutator {
+ public:
+  explicit ToMixedPrecisionRewriter(const VarDTypeMap* var_dtype_map, DataType output_dtype)
+      : var_dtype_map_(var_dtype_map), output_dtype_(output_dtype) {}
+
+ private:
+  NTensorStruct NTensorStructToRealize(const NTensorStruct& cur, const DTypeSet& dtype_set) {
+    // For each leaf: if only fp32 is required, realize fp32, otherwise realize fp16
+    if (cur.IsLeaf()) {
+      bool all_fp32 = true;
+      for (const NTensorStruct& dtype : dtype_set) {
```

Comment:
Add a variant of MapToNestedMsg that takes NestedMsg<X> and return NestedMsg<Y>

Reply:
It doesn't apply here since dtype_set needs handling when we recurse into fields